### ISSUE TITLE
Release location moved, crashing on recompile

### ISSUE THREAD

**Issue by choptastic** [rebar/rebar#503]
If a reltool-generated release is moved (that is, if the entire directory structure is moved to a different location), running rebar from within that release (and using the release's Erlang version), it crashes.

The full error message is:

```erlang
ERROR: compile failed while processing /home/gumm/www/nitrogen-git/myapp/lib/mimetypes: {'EXIT',{{badmatch,{error,enoent}},
         [{rebar_erlc_compiler,modify_erlcinfo,4,
                               [{file,"src/rebar_erlc_compiler.erl"},
                                {line,443}]},
          {rebar_erlc_compiler,'-update_erlcinfo_fun/2-fun-0-',4,
                               [{file,"src/rebar_erlc_compiler.erl"},
                                {line,399}]},
          {lists,foldl,3,[{file,"lists.erl"},{line,1261}]},
          {rebar_erlc_compiler,update_erlcinfo,3,
                               [{file,"src/rebar_erlc_compiler.erl"},
                                {line,418}]},
          {rebar_erlc_compiler,'-update_erlcinfo_fun/2-fun-0-',4,
                               [{file,"src/rebar_erlc_compiler.erl"},
                                {line,399}]},
          {lists,foldl,3,[{file,"lists.erl"},{line,1261}]},
          {rebar_erlc_compiler,init_erlcinfo,2,
                               [{file,"src/rebar_erlc_compiler.erl"},
                                {line,389}]},
          {rebar_erlc_compiler,doterl_compile,4,
                               [{file,"src/rebar_erlc_compiler.erl"},
                                {line,300}]}]}}
```

Upon further inspection is appears in trying to read the contents of a file that no longer exists:

In my case, the `Source` file generating the error is WARN: "/www/nitrogen-git/nitrogen/rel/nitrogen/lib/parsetools-2.0.12/include/leexinc.hrl"

This file was originally generated there, but the whole directory was moved to /www/nitrogen-git/myapp.  The naive approach is modifiying `rebar_erlc_compiler:modify_erlcinfo/4` to not crash if the file does not exist seems to work, but it feels like a bandaid fix and not a real fix (and not totally knowing what exactly that module is doing, I was hoping someone with a better internal model of that code would know where a better fix would be applied).

**Comment by tuncer** [comment]
@kejv, what do you think?

**Comment by tuncer** [comment]
Given that each project has its own `.rebar/erlcinfo`, we should store the relative path instead.

**Comment by ferd** [comment]
I think we can't properly store relative paths because rebar just `cd`s everywhere all the time, doesn't it? So if you're down in a dep, the path needs to be absolute to be properly readable from the project root later on, no?

**Comment by tuncer** [comment]
Each project has its own `.rebar/erlcinfo`, so relative paths are fine for project-local files. External files like the parsetools header are a problem, and updating the entry with the new absolute path seems like the solution there.

**Comment by tuncer** [comment]
I have another idea how to solve this. Let's see if it works out...

**Comment by tuncer** [comment]
@choptastic can you try the wip-include_lib branch in my fork, and see if it helps?

**Comment by tuncer** [comment]
@choptastic can you try the wip-include_lib branch in my fork, and see if it helps? If it works, I'd like to see a full log (-vv) to be sure.

**Comment by choptastic** [comment]
I'll check of out within the hour

**Comment by tuncer** [comment]
I'll add extra debug logs, please wait.

**Comment by tuncer** [comment]
Logs modified.

**Comment by tuncer** [comment]
@choptastic I need you to provide logs for three runs:

1. first run
2. second run, after moving stuff around
3. last run, rebuilding again without moving stuff around

**Comment by tuncer** [comment]
@choptastic I need you to provide logs for three runs:

1. first run
2. second run, after moving stuff around
3. last run, rebuilding again without moving stuff around

It's probably best to put them in a gist.

**Comment by tuncer** [comment]
Signing out for today.

**Comment by choptastic** [comment]
Here's one big gist, captured with `script` and lightly annotated to differentiate each run.

As I understand your request, here's what I did, in short.
1) Build the nitrogen so that it's not moved around at all (the default nitrogen behavior is to move a new release to a different directory, so specifying the PROJECT and PREFIX locations as I did ensure they stay where they are)
2) Move the release to a new location on the filesystem: (`mv rel/nitrogen ../myapp`), `cd` into that dir, and run `make` (which just runs `./rebar get-deps compile`)
3) Move the release back to its original location and run it again, demonstrating that it works in the original location.

https://gist.github.com/choptastic/e0fe1705ae3c1b815f2a

**Comment by choptastic** [comment]
Also, I was a little wrong, since the app gets built and run from the released as a self-contained thing, the actual commands as executed from the Makefile looks more like this:

```make
compile:
       @(export PATH=`pwd`/`echo erts-*/bin`:$$PATH; echo "Using Erlang in `which erl`"; $(REBAR) compile)
```
Which basically ends up running as:

```export PATH=/www/nitrogen-git/nitrogen/erts-6.3/bin:$PATH; ./rebar -vv compile```

This is to force the release to always use the release-copied version of Erlang.

**Comment by tuncer** [comment]
I cannot see any include_lib message during `==> mimetypes (compile)`. Any idea?

Also, `mimetypes/rebar.config` seems to contain `{erl_opts, [debug_info, fail_on_warning]}`, but `debug_info` is enabled by default and `fail_on_warning` must be `warnings_as_errors`. So the correct option is `{erl_opts, [warnings_as_errors]}`.

**Comment by choptastic** [comment]
> I cannot see any include_lib message during ==> mimetypes (compile). Any idea?

Perhaps this can help.  Mimetypes doesn't use include_lib to include those headers, it uses a -file() directive, (I'm not sure what that is), and it's in a generated mimetypes_parse.erl file.

This is the mimetypes_scan.erl file that gets generated: https://gist.github.com/choptastic/d6fbec543e796c528a8e

And I'd guess it's crashing on this, which apparently gets hardcoded into the source code:
-file("/www/nitrogen-git/nitrogen/rel/nitrogen/lib/parsetools-2.0.12/include/leexinc.hrl", 14).

Also, thanks for the fixes for the rebar.config.

**Comment by tuncer** [comment]
Okay, I see. You wrote

> In my case, the Source file generating the error is WARN: "/www/nitrogen-git/nitrogen/rel/nitrogen/lib/parsetools-2.0.12/include/leexinc.hrl"

and from that it wasn't evident this is a -file() attribute. So, it's actually leex embedding attributes with absolute paths. From what I can tell, the same problem exists with yecc.

The logical solution seems to regenerate mimetypes_parse.erl if the absolute path becomes invalid.

We could try to detect a leex-generated file and manipulate the paths, but that's doesn't seem like rebar's responsibility. The same error would happen if you run `erlc mimetypes_parse.xrl`.

**Comment by tuncer** [comment]
cc @rvirding @ferd

**Comment by tuncer** [comment]
Nevertheless, can you please use the scenario to test the changes in my branch by adding an include_lib("kernel/include/file.hrl") to one of the files and repeat the steps (with -vv logs). That's because the bug which exists (but doesn't affect you yet) is real, and the branch should fix it.

**Comment by choptastic** [comment]
> from that it wasn't evident this is a -file() attribute. So, it's actually leex embedding attributes with absolute paths. From what I can tell, the same problem exists with yecc.

Agreed. I honestly didn't know what was causing it. It only first appeared after fixing #350

> We could try to detect a leex-generated file and manipulate the paths, but that's doesn't seem like rebar's responsibility. The same error would happen if you run erlc mimetypes_parse.xrl.

The thing that strikes me here is that it works just fine with rebar 2.5.0. I haven't dug into any of the details about what the differences are between 2.5.0 and HEAD that would cause it to break.

> Nevertheless, can you please use the scenario to test the changes in my branch by adding an include_lib("kernel/include/file.hrl") to one of the files and repeat the steps (with -vv logs). That's because the bug which exists (but doesn't affect you yet) is real, and the branch should fix it.

Sure, I'll give it a check in a bit.

**Comment by tuncer** [comment]
> The thing that strikes me here is that it works just fine with rebar 2.5.0. I haven't dug into any of the details about what the differences are between 2.5.0 and HEAD that would cause it to break.

That's because of the new dependency handling code. We process -file() attributes as dependencies of that source file, and the problem is that the path is absolute.

We can experiment with modifying the leex:file() call in rebar_erlc_compiler with a custom {includefile, "path_to_/leexinc.hrl"}:

1. You could try a relative path, but a directory move can break that as well. 
2. You can try but I doubt -file() can be made to work like -include_lib() (which resolves the path via code:lib_dir()).
3. rebar could cache a copy of leex's and yecc's prologue header locally and always use a relative path via {includefile, ".rebar/leexinc.hrl"}

After doing the tests for the include_lib bug, manually cache leexinc.hrl somewhere local in mimetypes/ and modify the leex:file call in rebar to reference with a relative path. If that works, it's something we could automate in rebar, but I'm not sure we should.

**Comment by rvirding** [comment]
Sorry missed the initial discussion. Leex adds these so that compiler
errors give the right file name, the use file name, and line numbers. Yecc
does the same. I don't really think it is that much of a need to do this
but they do.

Don't know how you would fix it though. What do you do with yecc generated
files?

Robert

**Comment by tuncer** [comment]
> Sorry missed the initial discussion. Leex adds these so that compiler errors give the right file name, the use file name, and line numbers. Yecc does the same. I don't really think it is that much of a need to do this but they do.
>
> Don't know how you would fix it though. What do you do with yecc generated files?

rebar takes -file() attributes into account when looking for a source file's dependencies to use for the needs-(re)compile check.

What happens here is that original location of the leexinc.hrl changes in a Nitrogren-based project after the initial generation of the parser from .xrl.

This crashes in rebar before the compile:file call because we fail to check the now missing leexinc.hrl. @choptastic any idea how this didn't crash before (old rebar) due to the missing file? What happens if you try to compile the file with erlc (after leexincl.hrl's is missing)?

**Comment by tuncer** [comment]
@choptastic are you sure 2.5.0 doesn't crash? The same -file() handling code is present there as well.

**Comment by tuncer** [comment]
> Sorry missed the initial discussion. Leex adds these so that compiler errors give the right file name, the use file name, and line numbers. Yecc does the same. I don't really think it is that much of a need to do this but they do.
>
> Don't know how you would fix it though. What do you do with yecc generated files?

rebar takes -file() attributes into account when looking for a source file's dependencies to use for the needs-(re)compile check.

What happens here is that original location of the leexinc.hrl changes in a Nitrogren-based project after the initial generation of the parser from .xrl.

This crashes in rebar before the compile:file call because we fail to file:open the now missing leexinc.hrl.

**Comment by choptastic** [comment]
@tuncer, it does indeed compile fine with 2.5.0.  Here's the output of building it with rebar 2.5.0 after moving it away from its original location: https://gist.github.com/choptastic/52e04b6ec634e8e849b4

**Comment by tuncer** [comment]
@choptastic are you sure mimetypes is compiled at least twice? I can only see one `==> mimetypes (compile)`.

**Comment by tuncer** [comment]
To be clear, the old -file() attribute should fail, if you generated the file (compiled the project) before the dir-move.

**Comment by choptastic** [comment]
In the script above there, it had already been compiled in the original location, hence why only 1 compile notification.

As far as compiling with erlc, it does compile even with the broken hard paths that don't exist:  Here is a relevant log of the console showing the file with the broken path, and it recompiling with erlc.

https://gist.github.com/choptastic/3c650b309d7b130d9296

**Comment by tuncer** [comment]
erlc seems to ignore -file() attributes referencing non-existing files. We should do the same and log a DEBUG msg.

**Comment by tuncer** [comment]
Please try the ignore-invalid-file-attr branch and post a -vv log.

**Comment by choptastic** [comment]
Here's the run above with the include_lib patch (mimetype removed so it would succeed). There were no crashes.

I added `-include_lib("kernel/file/include.hrl").` to site/src/index.erl

https://gist.github.com/choptastic/bda8deea55eb6ab3894a

**Comment by choptastic** [comment]
Touched site/src/index.erl (using include_lib patch):

https://gist.github.com/choptastic/0572e046f42bee285bef

**Comment by choptastic** [comment]
Using the ignore-invalid-file-attr branch, second run (after moving file locations):

https://gist.github.com/choptastic/d4cad2896bf2b3603eaa

**Comment by tuncer** [comment]
The file gets cached in nitrogen/.rebar/erlcinfo and is expected to still exist after stuff is moved around. That's the failing file:open/2 call.

**Comment by tuncer** [comment]
Please try the erlc-unstable-deps branch.

**Comment by tuncer** [comment]
Please wait, there's other bugs.

**Comment by tuncer** [comment]
I need to incorporate more fixes. As of yet, unfinished...

**Comment by tuncer** [comment]
Please try my erlc-unstable-deps branch. It's full of debug logs, and the re-expanded path is not updated in the digraph, but it should fix the incorrect file:open/2 call. I'll probably have a look at the digraph update problem on Sunday.

**Comment by tuncer** [comment]
The current iteration of erlc-unstable-deps has the right idea, but the implementation is flawed. Back to the drawing board.

**Comment by tuncer** [comment]
@choptastic, can you please test my dep-open-crash branch?

**Comment by choptastic** [comment]
Confirmed to be working on initial run. I will produce a -vv output for you in a moment.

**Comment by tuncer** [comment]
And current master does not work, right?

**Comment by choptastic** [comment]
Initial build which builds the release and compiles, then copies to new directory: http://paste.ubuntu.com/11631818/
Second build which tries to run from the new location: http://paste.ubuntu.com/11631938/

And confirmed to be broken in master.

**Comment by choptastic** [comment]
I know you've been stressing about this, so I want you to know I appreciate you fixing this, even if this isn't quite the complete fix you want (unless it is, in which case, awesome).

The following is also true: The release is built with the current master version of rebar (built, but subsequent `make`s from other directories fail). Then the rebar binary is replaced with your patched version. Then run `make` again with your new binary (which succeeds). Then copying the old (that is, master) version of rebar back into the project. Then running `make` succeeds. So your rebar must be cleaning up the erlcinfo file so that the old rebar still works.

Either way, it's building with your patch, so kudos for your hard work.

**Comment by tuncer** [comment]
It's not what I aimed to achieve, but a simple fix which makes sure that we treat the error case correctly.

There was a missing newline in the WARN message, which I just fixed in the branch, so please update your rebar.

Also, what's the log after re-running the test (with the new rebar)? If the 3rd and 4th logs are good as, this should suffice for now.

**Comment by tuncer** [comment]
pull request: #509

**Comment by tuncer** [comment]
Thanks for the bug report and testing.

**Comment by choptastic** [comment]
Of course. Happy to help.

I'll get the report with the new change for you a bit later today, as I'm
currently away from the computer.

**Comment by david-kubecka** [comment]
Hi, I have just got into hits. @tuncer Please let me think for a minute about it before merging the pull request.

**Comment by david-kubecka** [comment]
After reading through this issue it is still puzzles me how it can happen. Because when the assumed root cause is that erlcinfo is copied to new release location which unchanged node/file names, then I don't know how `modify_erlcinfo/4` can be called without `Source` being an existing file. This can only happen when `digraph:vertex(G, Source)` in `update_erlcinfo/3` returns false which means that `Source` doesn't come from the original graph.

I think the problem is somehow connected to this part of the compile log:
```
WARN:  File missing. Remove from dep graph:
/www/nitrogen-git/nitrogen/rel/nitrogen/lib/parsetools-2.0.12/include/leexinc.hrl
WARN:  Failed to open file: /www/nitrogen-git/nitrogen/rel/nitrogen/lib/parsetools-2.0.12/include/leexinc.hrl
{error,enoent}WARN:  Failed to open file: /www/nitrogen-git/nitrogen/rel/nitrogen/lib/parsetools-2.0.12/include/leexinc.hrl
{error,enoent}WARN:  File missing. Remove from dep graph:
```
Here it seems as if the leexinc.hrl file was added (and removed) several times to (and from) the dep graph which is quite strange.

**Comment by tuncer** [comment]
@david-kubecka, I agree that modify_erlcinfo() shouldn't have to deal with this at all. To be clear, the patch is for unblocking the 2.6.0 release, as I ran out of time for implementing a proper fix. That's why I filed #506. If you agree this is fine to merge, we can discuss the proper solution there, and get the already belated release out the door.

**Comment by kejv** [comment]
@tuncer I understand (previously was writing from my company github account, sorry). The patch seems ok. I will copy my investigation to #506 so that it doesn't get lost.

**Comment by tuncer** [comment]
@ferd just merged #509.

**Comment by tuncer** [comment]
This can be closed.
